mkosi/__init__.py

```diff
@@ -351,47 +345,71 @@ def mount_build_overlay(context: Context, volatile: bool = False) -> Iterator[Pa
 
 
 @contextlib.contextmanager
-def finalize_scripts(scripts: Mapping[str, Sequence[PathString]] = {}) -> Iterator[Path]:
+def finalize_scripts(scripts: Mapping[str, Sequence[PathString]], root: Path) -> Iterator[Path]:
     with tempfile.TemporaryDirectory(prefix="mkosi-scripts") as d:
+        # Make sure than when mkosi-as-caller is used the scripts can still be accessed.
+        os.chmod(d, 0o755)
+
         for name, script in scripts.items():
             # Make sure we don't end up in a recursive loop when we name a script after the binary it execs
             # by removing the scripts directory from the PATH when we execute a script.
-            (Path(d) / name).write_text(
-                textwrap.dedent(
-                    f"""\
-                    #!/bin/sh
-                    DIR="$(cd "$(dirname "$0")" && pwd)"
-                    PATH="$(echo "$PATH" | tr ':' '\\n' | grep -v "$DIR" | tr '\\n' ':')"
-                    export PATH
-                    exec {shlex.join(str(s) for s in script)} "$@"
-                    """
-                )
-            )
+            with (Path(d) / name).open("w") as f:
+                f.write("#!/bin/sh\n")
+
+                if find_binary(name, root=root):
+                    f.write(
+                        textwrap.dedent(
+                            """\
+                            DIR="$(cd "$(dirname "$0")" && pwd)"
+                            PATH="$(echo "$PATH" | tr ':' '\\n' | grep -v "$DIR" | tr '\\n' ':')"
+                            export PATH
+                            """
+                        )
+                    )
+
+                f.write(f'exec {shlex.join(str(s) for s in script)} "$@"\n')
 
             make_executable(Path(d) / name)
+            os.chmod(Path(d) / name, 0o755)
             os.utime(Path(d) / name, (0, 0))
 
         yield Path(d)
 
 
+GIT_COMMAND = (
+    "git",
+    "-c",
+    "safe.directory=*",
+)
+
+
+def mkosi_as_caller() -> tuple[str, ...]:
+    return (
+        "setpriv",
+        f"--reuid={INVOKING_USER.uid}",
+        f"--regid={INVOKING_USER.gid}",
+        "--clear-groups",
+    )
+
+
 def finalize_host_scripts(
     context: Context,
-    helpers: Mapping[str, Sequence[PathString]],
+    helpers: Mapping[str, Sequence[PathString]] = {},
```

Comment:
Hmm, using a mutable object as the default is considered bad practice. I think this should be None or ().

Reply:
@keszybz The `Mapping` type is immutable. The reason using a mutable object is bad practice is because it could be mutated. However, the type is immutable here which means it cannot be mutated so it's fine. We use this pattern all over the place so let's not try to fix this here.